packages/nx/src/utils/provenance.ts

```diff
@@ -0,0 +1,95 @@
+import { execFile } from 'child_process';
+import { promisify } from 'util';
+
+export async function checkPackageHasProvenance(
+  packageName: string,
+  packageVersion: string
+): Promise<{
+  success: boolean;
+  error?: string;
+}> {
+  // this is used for locally released versions without provenance
+  // do not set this for other reasons or you might be exposed to security risks
+  if (process.env.NX_MIGRATE_SKIP_PROVENANCE_CHECK) {
+    return {
+      success: true,
+    };
+  }
+  const execFileAsync = promisify(execFile);
+
+  const npmViewResult = JSON.parse(
+    (
+      await execFileAsync(
+        'npm',
+        ['view', `${packageName}@${packageVersion}`, '--json'],
+        {
+          timeout: 20000,
+        }
+      )
+    ).stdout.trim()
+  );
+
+  const attURL = npmViewResult.dist?.attestations?.url;
+
+  if (!attURL) return { success: false, error: 'No attestation URL found' };
+
+  // refer to https://slsa.dev/spec/v1.1/provenance#schema for the shape of this object
+  const attestations = (await (await fetch(attURL)).json()) as any;
+
+  const provenanceAttestation = attestations?.attestations?.find(
+    (a) => a.predicateType === 'https://slsa.dev/provenance/v1'
+  );
+  if (!provenanceAttestation)
+    return { success: false, error: 'No provenance attestation found' };
+
+  const dsseEnvelopePayload = JSON.parse(
+    Buffer.from(
+      provenanceAttestation.bundle.dsseEnvelope.payload,
+      'base64'
+    ).toString()
+  );
+
+  const workflowParameters =
+    dsseEnvelopePayload?.predicate?.buildDefinition?.externalParameters
+      ?.workflow;
+
+  // verify that provenance was actually generated from the right publishing workflow
+  if (workflowParameters?.repository !== 'https://github.com/nrwl/nx') {
+    return { success: false, error: 'Repository does not match nrwl/nx' };
+  }
+  if (workflowParameters?.path !== '.github/workflows/publish.yml') {
+    return {
+      success: false,
+      error: 'Publishing workflow does not match .github/workflows/publish.yml',
```

Comment:
This would be more idiomatic if it threw an Error vs not throw an Error (indicating success).

Also, I think the error messages should show remediation steps. Maybe double-check if the version is actually proper. And also show that this check can be skipped so users have a way around it if for some reason the version if proper.

I also know.. sometimes I visit npmjs.com and it has trouble fetching provenance? 🤔 If the attestation servers are down, users might want to skip this 🤷

Reply:
yes, I updated it to throw an error, it's cleaner.
Also added an instruction to check the package at npmjs.org.

I don't know about the last bit, I would be very hesitant about adding an instruction to skip here as people don't really read error messages and if they don't know what provenance is, they might just ignore the warning and skip the check.